node/src/reactor/initializer.rs

```diff
@@ -263,6 +259,29 @@ impl Reactor {
     }
 }
 
+fn check_integrity(state_roots: &[Digest], missing_trie_keys: &[Digest]) -> Result<(), Error> {
+    if missing_trie_keys.is_empty() {
+        // No missing trie keys; integrity ok.
+        return Ok(());
+    }
+
+    // There are some missing trie keys present. Check if they are all
+    // state root hashes.
+    let mut missing_trie_keys = missing_trie_keys.to_vec();
+    missing_trie_keys.retain(|missing_trie_key| !state_roots.contains(missing_trie_key));
```

Comment:
`state_roots` could contain hundreds of thousands of hashes (one per block), so we might want to optimize this part. I'd suggest collecting the hashes into a `HashSet` first, and use that `HashSet` for the `contains` check.